text/0002-signature-based-completions.md

```diff
@@ -0,0 +1,216 @@
+- Feature Name: signature-based-completions
+- Start Date: 2020-07-25
+- RFC PR: [nushell/rfcs#0002](https://github.com/nushell/rfcs/pull/2)
+- Nushell Issue: [nushell/nushell#0000](https://github.com/nushell/nushell/issues/0000)
+
+# Summary
+
+Add a completion engine to nushell that can determine how to complete an item at the cursor by
+parsing the line, extracting a signature, and using the information in that signature to determine
+what kind of thing needs to be completed.
+
+# Motivation
+
+We currently use rustyline's completer to do our completions. This completer is unfortunately
+limited, as it focuses on filenames only. We've tried to expand this with command names, but it
+scatters completion code all over the codebase.
+
+We already have incredibly rich information in signatures for commands, and intend to eventually use
+this to describe external commands. By building our completions around signatures, we can take
+advantage of highly contextual information to make nushell's completion system highly contextual,
+powerful, and delightful to use.
+
+# Guide-level explanation
+
+Completion is the act of responding to a user request to find a list of recommendations that fit
+some incomplete information that exists at their cursor. For example, if the user presses tab at the
+end of this line:
+
+    > ls dir/
+
+all path entries under the subdirectory `dir` could be offered as suggestions. Similarly, for
+
+    > ls dir/a
```

Comment:
Are you homogenizing path separators here so slash will be the path separator for windows/mac/linux? Or is this just a linux/mac example?

Reply:
Yeah, this is just a linux/mac example. I don't dive into too much detail about the specific completers themselves, but the filename/path "completer" would have to take that into consideration 💯